### ISSUE TITLE
Unexpected behavior on casting to bool

### ISSUE THREAD

**Issue by KonstantinKuklin** [php-ds/extension#62]
Lets try some simple magic:

``` php
$ds = \Ds\Vector();

Expected:
if(!$ds) {
  //here
}
```

Actual:

```
[Fatal Error] Object of class Ds\Vector could not be converted to boolean at
```

And another case is

``` php
$ds = \Ds\Vector([1]);

Expected:
if($ds) {
  //here
}
```

Actual:

```
[Fatal Error] Object of class Ds\Vector could not be converted to boolean at
```

So, is it a bug or feature?
Because I change array to vector and my code is not work, because if for array is ok, but not for Ds structures.

**Comment by redneck1** [comment]
Doing like if (!$myArr) is an awful practice. What does it mean to you to cast an array (or any other sort of collection) to bool? If it means empty collection, use if (0 === count($myArr)).

It's a bad feature that PHP allows you cast a standard PHP-array and some another as well...

Just try:

php -r 'if (!(new SplFixedArray())) {echo 0;} else {echo 1;}' // 1
php -r 'if (![]) {echo 0;} else {echo 1;}' // 0

What's logic then???

**Comment by nikic** [comment]
I agree that it's best not to overload the boolean cast operator. SimpleXMLElement is one of the few (maybe even the only?) object that overloads boolean casts, and this has caused a lot of grief. People rely on all objects being truthy in PHP and this would break the convention.

Btw, the Collection interface has an `isEmpty()` method, so you can write `if($foo->isEmpty())`.

**Comment by rtheunissen** [comment]
Use `isEmpty()`, which is better anyway because it communicates exactly what your conditional is checking.

**Comment by rvanvelzen** [comment]
You can't use `isEmpty()` reliably in this kind of case:

```php
$set = $this->optionalSet(); // returns `Set` or `null`
if (!$set) {
    return false;
}
```

**Comment by rtheunissen** [comment]
You'd have to use `! $set || $set->isEmpty()`.

**Comment by rvanvelzen** [comment]
That does not work:

```
>>> $set = null;
=> null
>>> ! $set || $set->isEmpty();
=> true
>>> $set = new \Ds\Set();
=> Ds\Set {#197}
>>> ! $set || $set->isEmpty();
PHP error:  Object of class Ds\Set could not be converted to boolean on line 1
```

Unless I'm missing something obvious?

**Comment by rtheunissen** [comment]
Ohh I see what you mean here. Will fix. 😊

**Comment by rtheunissen** [comment]
Fixed in 6006e411d8d9503a45b5e5011370cbceb2397c39 and released as [1.2.1](https://pecl.php.net/package/ds)

**Comment by rvanvelzen** [comment]
Thanks!
